src/librbd/api/Group.h

```diff
@@ -46,6 +49,8 @@ struct Group {
                          const char *old_snap_name, const char *new_snap_name);
   static int snap_list(librados::IoCtx& group_ioctx, const char *group_name,
                        std::vector<group_snap_info_t> *snaps);
+  static int snap_list2(librados::IoCtx& group_ioctx, const char *group_name,
+			                std::vector<cls::rbd::GroupSnapshot> *cls_snaps);
```

Comment:
The same comment about using `cls::rbd::GroupSnapshot` here. And we already have a method with `group_snap_info_t`. Why aren't you happy with it?

Reply:
I need to get the images snap id under group snap.
`group_snap_info_t` only has information about group snap.
If I remove `snap_list2()` and `group_image_list2()` interfaces,
I need to change the `group_snap_info_t` and `group_image_info_t `structures as follows:
```
typedef struct {
  std::string name;
+ std::string image_id;
  int64_t pool;
  group_image_state_t state;
} group_image_info_t;

+typedef struct {
+ std::string name;
+ std::string image_id;
+ uint64_t snap;
+ int64_t pool;
+} group_image_snap_info_t;

typedef struct {
  std::string name;
+ std::vector<group_image_snap_info_t> snaps;
  group_snap_state_t state;
} group_snap_info_t;
```

Reply:
In API, we can't add new fields to existing structures. We would need to introduce new structures instead and probably deprecate the old ones, which I am not convinced yet is a good idea just to solve this particular case.

I believe it possible to collect all necessary information with existing API. You will just need do more calls. E.g. you will have to have every image open, to make additional calls to get the image id and snapshot information. But I suppose you will have to open every image anyway, so it should not be a big problem.

So I would try to implement it with  existing API, although it will be longer and not so nice.  If some info is not accessible indeed, then provide a function. And think about new structures as the last resort.

Note, I also think that the current group API is not very nice and needs improvements. So eventually we will improve this, I am just not sure we have to do this just right now.

Reply:
Ok, I see what you mean, but there is one piece of informationthat I don't think I can get:
When incrementally exporting, I cannot get images information(image_id, snap_id, pool_id) from a group snapshot.
For example, I can't get the images info in group1@v1,  and I can't get the information of how many images there are under snap v1 and the image id,snap id,pool id of each image.

So I would like to add an API and  a structure as follows:
```
+typedef struct {
+  std::string image_id;
+  uint64_t snap;
+  int64_t pool;
+} group_image_snap_info_t;

+int RBD::group_snap_image_list(IoCtx& group_ioctx, const char *group_name,
+                           const char *group_snap_name,
+                           std::vector<group_image_snap_info_t> *images,
+                           size_t group_image_snap_info_t);
```

Reply:
LGTM. 

nit:  I would use `snap_id` and `pool_id`.

Reply:
delete the API: snap_list2

[https://github.com/ceph/ceph/pull/40808/commits/c058e51e97d50578a0e7c46b0bd66dd4f2e5db96](url)